### PR TITLE
Dependency: Migration to PyQt6

### LINKED ISSUE #1301: Migration to PyQt6 until the end of 2023 because of PyQt5 support end
Currently we are using `PyQt5`.
The original Qt5 (by Qt Company) [ends its support in the second quarter of 2023](https://www.qt.io/blog/qt-5.15-extended-support-for-subscription-license-holders#contactopen).
But different from original Qt the [support for PyQt5 ends until the end of 2023](https://www.riverbankcomputing.com/pipermail/pyqt/2022-September/044903.html) what buys us a little more time.

Our luck is that the qt code is quit isolated from the rest.
And [this blog post about migrating PyQt 5 to 6](https://www.pythonguis.com/faq/pyqt5-vs-pyqt6/) stated that it seems quit easy to migrate.

We also have to take into account the backwards compatibility to PyQt5. But this also seems to be quite easy. But maybe it doesn't matter anymore in the end of 2023.

Sidenote: What I knew but didn't realized in the first place. Qt is created by _Qt Company_. But `PyQt` is created by _Riverbank Computing_.

**EDIT**: For new contributors this is a good opportunity to help in that project without touching the code. Use the [branch qt6migration at buhtz/backintime](https://github.com/buhtz/backintime/tree/qt6migration) and play around with the GUI to find bugs introduced by this migration.

### LINKED ISSUE #1607: Manual testing of PyQt6 code
The migration (#1301) from PyQt5 to PyQt6 is ongoing and happens in a forked repo in a separate branch: [`buhtz/backintime:qt6migration`](https://github.com/buhtz/backintime/tree/qt6migration).

Because of that migration there is a high risk of introduced bugs because of name typos causing `AttributeError`. The code is not covered by unit tests. Therefore a manual testing is needed.

Install the PyQt6 version of Back In Time (using [`buhtz/backintime:qt6migration`](https://github.com/buhtz/backintime/tree/qt6migration) or a fork from it) and try to open and use all GUI components (dialogs, buttons, OK/Yes/No/Cancel, ...) available. Report the discovered errors or fix them and open a RP.

### PR DESCRIPTION
**Pull request by buhtz**
And here it comes... The Qt6 version of BIT.

No hurry with migration. Maybe doing some further manual tests.
The majority of modifications are just renaming some classes and enums because Qt moved some inside there library from one namespace to another. I am not aware of relevant modifications in the behavior or functionality between Qt5 and Qt6.

I did tests myself and found some minor problems and typos I was able to fix.

Close #1301
Close #1607

### CODE REVIEW

**Comment by aryoda** [comment]
I haven an open issue (#1630) for the most-recent release (and more may always arrive).

How shall we organize to work in parallel to fix high-bugs but also developing the Qt6 version.

Since distro pkg maintainer may rely on releases as tags on our `dev` branch merging the Qt6 version into dev too early
may cause extra work for them to (temporarily) use another branch for the fixes.

**Comment by buhtz** [comment]
No need to merge Qt6 so early. So I set it to Draft. Just keep the PR open. I do sync it with "dev" from time to time.
It is how it worked the last 2 months.

**Comment by aryoda** [comment]
Perfect, so changes and fixes regarding Qt6 go to this PR (= your repo) until we merge it into `dev`.

**Comment by buhtz** [comment]
How is the current status?

I did a manual systematic check of the GUI clicking on every button in every dialog in every of our 4 snapshot modes. I also use pylint to find relevant code smells.

~~I found Line 74 in serviceHelper.py.
I am able to `import dbus.mainloop` but not `import dbus.mainloop.pyqt5` (yes: 5). So despite our Qt6 migration why doesn't this trigger an exception on my system? When is this serverHelper.py used?~~
No problem. A service I do not use. Might be udev related. And the dependency (python3-dbus.mainloop.pyqt5) is listed in our CONTRIBUTION.md. My fault. Everything is fine.

In the end I am not able to find any problems anymore.

**Comment by aryoda** [comment]
> How is the current status?

Do you mean the testing status?

TBH: Still tbd :-(

I think I can manual testing this Sunday (I hope - depends on the weather ;-)

**Comment by buhtz** [comment]
Really, no hurry. I just wasn't clear about the next step. Take your weekend.

File: qt/app.py [hunk 1]
```diff
@@ -719,11 +720,11 @@ def _create_main_toolbar(self):
             # Note that Qt use "text()" as "toolTip()" by default.
             if act.toolTip() != act.text():
 
-                if toolbar.layoutDirection() == Qt.LayoutDirection.RightToLeft:
+                if QApplication.instance().isRightToLeft():
```
Comment (aryoda) [inline qt/app.py:723]: Checking for right-to-left and taking left-to-right as "else" part is a sensible "defensive" way (already was that way before the change), so look fine...

**Comment by aryoda** [comment]
is this PR "stable" for testing (I wanna do some manual tests tonight)?

**Comment by buhtz** [comment]
> is this PR "stable" for testing (I wanna do some manual tests tonight)?

Yup. It is waiting for you. ;)

**Comment by aryoda** [comment]
OK, I have cloned your repo and it is very annoying that Ubuntu 22.04 does not provide `python3-pyqt6`.

## TLDR ;-)

Packaging of our `PyQt6` BiT version for Ubuntu 22.04 will be a challenge for the package maintainer
since `PyQt6` is not contained in the official repo.

## Installation details for Ubuntu 22.04

Installation steps
- I have installed `qt6-base-dev` (not sure if this was really required)
- `pip3 install PyQt6` said
   > Successfully installed PyQt6-6.6.1 PyQt6-Qt6-6.6.2 PyQt6-sip-13.6.0
   > [...] into the user site directory `~/.local/lib/python3.10/site-packages/`
- The GUI did not start failing with `ModuleNotFoundError: No module named 'PyQt6'` (despite it was installed).
  I have removed `-Es` from the `python3` command line in `qt/backintime-qt` since this caused ignoring my `PyQt6` installed in the user site directory (even though this may cause unknown side effects on my tests because other packages may be loaded now)
- `qt/backintime-qt` does still not start the GUI with
  > qt.qpa.plugin: From 6.5.0, xcb-cursor0 or libxcb-cursor0 is needed to load the Qt xcb platform plugin.
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
- Installing the distro package `xcb-cursor0` solved this problem and the GUI started with `PyQt6`

Note: The distro packages `libglu1-mesa-dev` and  `libx11-xcb-dev` need possibly also to be installed according to [this post](https://askubuntu.com/questions/1485442/issue-with-installing-pyqt6-on-ubuntu-22-04) (in my case the packages were already installed).

## Unit tests

- `common` unit tests succeeded as good as for the Qt5 version (365 passed, 2 skipped) - not unexpected since this non-GUI code was not changed...


## Next steps

I am doing my manual GUI tests now...

**Comment by buhtz** [comment]
Ubuntu 22.04 LTS It is the second last (vorletztes) LTS.
24.04 is the latest LTS.
We are not responsible if companies support multiple LTS versions, especially not when it comes to Canonical.

The BIT version in Ubuntu 22.04 LTS is "1.2.1" and even Canonical won't change this. That is why it is LTS. I see no problem.

It is similar when it comes to PPAs. This are unsupported and 3rd party sources. Not our responsibility.

If users on 22.04 LTS need a fresh BIT they can install directly from upstream using our make system and PyPi for missing packages.

**Comment by aryoda** [comment]
Looks really great, I have done a lot of manual tests of the GUI and everything is working with `PyQt6` - what a smooth migration!

@buhtz  Thanks for your initiative and the work done to prepare the migration!

How could we proceed now?

BTW: What do you mean with "breaking change" in the `CHANGELOG`?
   The functionalities and public APIs are the same and changing dependencies we already did in the past.
   So I would prefix this with "Dependency" instead.
   The version increment from 1.4.x to 1.5.x is fine though IMHO (even if it is not "breaking" in the sense of no. 7 of the [semantic versioning guidelines](https://semver.org/#spec-item-7) but for us it is major milestone).

**Comment by buhtz** [comment]
I have to correct myself about Ubuntu. The 24.04 LTS will get released on 25th April 2024.

If there are users trying to install upstream-BIT on Ubuntu 22.* with PyQt6 installed from PyPi we could add a FAQ entry or something similar how they can hack it. Or modify the entry-point script checking the current Ubuntu release and removing the -Ee or something like this.

I do care about Ubuntu users of course but not about Canonical. :smile: 

> How could we proceed now?

Merge and shiver. :rofl: 

I do have some qt6-based PRs in the back just waiting for this merge.

> BTW: What do you mean with "breaking change" in the `CHANGELOG`?

You are absolutely right about that. I was just scared because so many files changed that I may have introduced a bug. Makes no sense of course.

**Comment by buhtz** [comment]
Dear @Germar , @graysky2 , @jean-christophe-manciot and @Fantu ,
please be aware that BIT now migrate from PyQt5 to PyQt6.

**Comment by buhtz** [comment]
> Looks really great, I have done a lot of manual tests of the GUI and everything is working with `PyQt6` - what a smooth migration!

btw: Was that an "Approve"? 😃

**Comment by aryoda** [comment]
> > Looks really great, I have done a lot of manual tests of the GUI and everything is working with `PyQt6` - what a smooth migration!
> 
> btw: Was that an "Approve"? 😃

Ah yes, forgot to click the "approave button"!

**Review by aryoda — APPROVED** [review]

**Comment by aryoda** [comment]
Side note: The version number is still 1.4.4-dev - shall we update this now to 1.5.0-dev or when releasing?

**Comment by buhtz** [comment]
> Side note: The version number is still 1.4.4-dev - shall we update this now to 1.5.0-dev or when releasing?

Now I am confused. 🤣 
But wasn't it the result of our last discussion that we always increase the dev version. I mean "1.4.4-dev" will never become "1.4.4" but "1.4.5" or "1.5.0". That is why we released "1.4.3" after "1.4.1". "1.4.2" never existed but its dev-version "1.4.2-dev".
As I explained in that past discussion I prefer the process without over jumping a version ("1.4.2-dev" would become "1.4.2"). But we decided to make it the other way ("1.4.2-dev" becomes "1.4.3").
Or did I misunderstand something?

**Comment by aryoda** [comment]
> Now I am confused. 🤣

Ahm, I confused something, sorry. 1.4.4-dev is OK (and already incremented for dev).

1.5.0 will be the release version.

BTW: Regarding the "gaps" in release version numbers (if eg. 1.4.2-dev becomes 1.4.2 after 1.4.2)
I could not even find a consistent opinion or advice in the semver issues (so its up to us):

- https://github.com/semver/semver/issues/942
- https://github.com/semver/semver/issues/841
- https://github.com/semver/semver/issues/124#issuecomment-20980208
- https://github.com/semver/semver/issues/442
